[DOC] Add kv processor content gap doc

vagimeli opened the pull request:
### Description
Adds documentation about kv processor; addresses content gap

### Issues Resolved
#5651 

### Checklist
- [X] By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and subject to the [Developers Certificate of Origin](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

vagimeli commented:
@reta @gaobinlong Please take a look at the draft kv processor documentation. We particularly need your input for a real-life example of a kv processor ingest pipeline (see the section Using the processor steps 1-3). Thank you for your review! 

For reference, if you'd like to see recent documentation for other processors, go to https://opensearch.org/docs/latest/ingest-pipelines/processors/index-processors/.

reta commented on _ingest-pipelines/processors/kv.md:
```diff
@@ -0,0 +1,118 @@
+
+# KV processor
+
+The `kv` processor automatically extracts specific event fields or messages that are `key=value` format. This structured format organizes your data by grouping it together based on keys and values. It's helpful for analyzing, visualizing, and using data such as user behavior analytics, performance optimizations, or security investigations. 
```
```suggestion
The `kv` processor automatically extracts specific event fields or messages that are  in a`key=value` format. This structured format organizes your data by grouping it together based on keys and values. It's helpful for analyzing, visualizing, and using data such as user behavior analytics, performance optimizations, or security investigations. 
```

reta commented:
> We particularly need your input for a real-life example of a kv processor ingest pipeline (see the section Using the processor steps 1-3).

@gaobinlong do you mind providing those, I believe you should have them ready, thank you

gaobinlong commented on _ingest-pipelines/processors/kv.md:
````diff
@@ -0,0 +1,118 @@
+```json
+POST _ingest/pipeline/kv-pipeline/_simulate
+
+<insert example from SME>
````
Here is:
```
{
  "docs": [
    {
      "_index": "testindex1",
      "_id": "1",
      "_source":{
         "message": "goodbye=everybody hello=world"
      }
    }
  ]
}
```

gaobinlong commented on _ingest-pipelines/processors/kv.md:
````diff
@@ -0,0 +1,118 @@
+
+```json
+
+<insert reponse example from SME>
````
Here is:
```
{
  "docs": [
    {
      "doc": {
        "_index": "testindex1",
        "_id": "1",
        "_source": {
          "hello": "world",
          "message": "goodbye=everybody hello=world",
          "goodbye": "everybody"
        },
        "_ingest": {
          "timestamp": "2023-12-06T09:59:21.823292Z"
        }
      }
    }
  ]
}
```

gaobinlong commented on _ingest-pipelines/processors/kv.md:
````diff
@@ -0,0 +1,118 @@
+
+```json
+PUT testindex1/_doc/1?pipeline=kv-pipeline
+<insert example from SME>
````
Here is:
```
{
  "message": "goodbye=everybody hello=world"
}
```

vagimeli commented:
@reta @gaobinlong Thank you for your review. I've addressed your feedback. I'll move this forward for doc and editorial review and then get the PR merged. Thank you!

vagimeli commented on _ingest-pipelines/processors/kv.md:
```diff
@@ -0,0 +1,147 @@
+
+The following table lists the required and optional parameters for the `lowercase` processor.
+
+| Name  | Required/Optional  | Description  |
```
We are updating the tables to now read Required/Optional. The first round originally said Required, but those need updating. I'll sync them this week. Good comment. Thanks!

vagimeli commented:
@hdhalter I accepted your suggestions. I'll update the first round of processors that used Required instead of Required/Optional and Name instead of Parameter. Thanks for the review.

natebower commented on _ingest-pipelines/processors/kv.md:
```diff
@@ -0,0 +1,147 @@
+
+# KV processor
+
+The `kv` processor automatically extracts specific event fields or messages that are in a`key=value` format. This structured format organizes your data by grouping it together based on keys and values. It's helpful for analyzing, visualizing, and using data such as user behavior analytics, performance optimizations, or security investigations. 
```
```suggestion
The `kv` processor automatically extracts specific event fields or messages that are in a `key=value` format. This structured format organizes your data by grouping it together based on keys and values. It's helpful for analyzing, visualizing, and using data such as user behavior analytics, performance optimizations, or security investigations. 
```

natebower replied:
Last sentence: I suggest just "It's helpful in use cases such as user behavior analytics...". Otherwise, "...and using data, such as user behavior analytics, performance optimization, or security investigation data."

natebower commented on _ingest-pipelines/processors/kv.md:
```diff
@@ -0,0 +1,147 @@
+The `kv` processor automatically extracts specific event fields or messages that are in a`key=value` format. This structured format organizes your data by grouping it together based on keys and values. It's helpful for analyzing, visualizing, and using data such as user behavior analytics, performance optimizations, or security investigations. 
+
+## Example
+The following is the syntax for the `kv` processor: 
```
Please add a line between the heading and the following text.

natebower commented on _ingest-pipelines/processors/kv.md:
```diff
@@ -0,0 +1,147 @@
+
+| Parameter  | Required/Optional  | Description  |
+|---|---|---|
+`field`  | Required  | The name of the field that contains the data to be parsed. Supports [template snippets]({{site.url}}{{site.baseurl}}/ingest-pipelines/create-ingest/#template-snippets). |
```
```suggestion
`field`  | Required  | The name of the field containing the data to be parsed. Supports [template snippets]({{site.url}}{{site.baseurl}}/ingest-pipelines/create-ingest/#template-snippets). |
```

natebower commented on _ingest-pipelines/processors/kv.md:
```diff
@@ -0,0 +1,147 @@
+`exclude_keys` | Optional | The keys to exclude from the document. Default is `null`. |
+`include_keys` | Optional | The keys for filtering and inserting. Default is to include all keys. |
+`prefix` | Optional | The prefix to add to the extracted keys. Default is `null`. |
+`strip_brackets` | Optional | If set to `true`, strips brackets `()`, `<>,` or `[]` and quotes `'` or `"` from extracted values. Default is `false`.
```
Can we put parentheses around the examples?: (`()`, `<>,` or `[]`) and quotes (`'` or `"`)

natebower commented on _ingest-pipelines/processors/kv.md:
```diff
@@ -0,0 +1,147 @@
+`trim_key` | Optional | The string of characters to trim from the extracted keys. | 
+`trim value` | Optional | The string of characters to trim from the extracted values. |
+`description`  | Optional  | A brief description of the processor.  |
+`if` | Optional | A condition for running this processor. |
```
```suggestion
`if` | Optional | A condition for running the processor. |
```

natebower commented on _ingest-pipelines/processors/kv.md:
```diff
@@ -0,0 +1,147 @@
+`if` | Optional | A condition for running this processor. |
+`ignore_failure` | Optional | If set to `true`, failures are ignored. Default is `false`. |
+`on_failure` | Optional | A list of processors to run if the processor fails. |
+`ignore_missing`  | Optional  | Specifies whether the processor should ignore documents that do not have the specified field. Default is `false`.  |
```
```suggestion
`ignore_missing`  | Optional  | Specifies whether the processor should ignore documents that do not contain the specified field. Default is `false`.  |
```

natebower commented on _ingest-pipelines/processors/kv.md:
```diff
@@ -0,0 +1,147 @@
+`ignore_failure` | Optional | If set to `true`, failures are ignored. Default is `false`. |
+`on_failure` | Optional | A list of processors to run if the processor fails. |
+`ignore_missing`  | Optional  | Specifies whether the processor should ignore documents that do not have the specified field. Default is `false`.  |
+`tag` | Optional | An identifier tag for the processor. Useful for debugging to distinguish between processors of the same type. |
```
Either "in order to distinguish" or "when distinguishing".

natebower commented on _ingest-pipelines/processors/kv.md:
```diff
@@ -0,0 +1,147 @@
+
+Follow these steps to use the processor in a pipeline.
+
+**Step 1: Create a pipeline.** 
```
```suggestion
**Step 1: Create a pipeline**
```

natebower commented on _ingest-pipelines/processors/kv.md:
```diff
@@ -0,0 +1,147 @@
+
+**Step 1: Create a pipeline.** 
+
+The following query creates a pipeline, named `kv-pipeline`, that uses the `kv` processor to extract  the `message` field of a document:
```
```suggestion
The following query creates a pipeline, named `kv-pipeline`, that uses the `kv` processor to extract the `message` field of a document:
```

natebower commented on _ingest-pipelines/processors/kv.md:
````diff
@@ -0,0 +1,147 @@
+```
+{% include copy-curl.html %}
+
+**Step 2 (Optional): Test the pipeline.**
````
```suggestion
**Step 2 (Optional): Test the pipeline**
```

natebower commented on _ingest-pipelines/processors/kv.md:
````diff
@@ -0,0 +1,147 @@
+```
+{% include copy-curl.html %}
+
+#### Response
````
Confirm heading level.

natebower commented on _ingest-pipelines/processors/kv.md:
````diff
@@ -0,0 +1,147 @@
+}
+```
+
+**Step 3: Ingest a document.**
````
```suggestion
**Step 3: Ingest a document**
```

natebower commented on _ingest-pipelines/processors/kv.md:
````diff
@@ -0,0 +1,147 @@
+```
+{% include copy-curl.html %}
+
+**Step 4 (Optional): Retrieve the document.**
````
```suggestion
**Step 4 (Optional): Retrieve the document**
```

natebower commented on _ingest-pipelines/processors/lowercase.md:
```diff
@@ -29,9 +29,9 @@ The following is the syntax for the `lowercase` processor:
+| Parameter  | Required  | Description  |
 |---|---|---|
-`field`  | Required  | The name of the field that contains the data to be converted. Supports template snippets. |
+`field`  | Required  | The name of the field that contains the data to be converted. Supports [template snippets]({{site.url}}{{site.baseurl}}/ingest-pipelines/create-ingest/#template-snippets). |
```
```suggestion
`field`  | Required  | The name of the field containing the data to be converted. Supports [template snippets]({{site.url}}{{site.baseurl}}/ingest-pipelines/create-ingest/#template-snippets). |
```

natebower reviewed (commented):
@vagimeli Please see my comments and changes and let me know if you have any questions. Thanks!
